library/modules/Items.cpp

```diff
@@ -678,14 +679,54 @@ string Items::getDescription(df::item *item, int type, bool decorate) {
     item->getItemDescription(&tmp, type);
 
     if (decorate) {
-        addQuality(tmp, item->getQuality());
+        // Special indicators get added in a specific order
+        // Innermost is at the top, and outermost is at the bottom
 
-        if (item->isImproved()) {
+        // First, figure out the quality levels we're going to display
+        int craftquality = item->getQuality();
+        int craftquality_only_imps = item->getImprovementQuality();
+        bool has_displayed_item_improvements = item->isImproved();
+        if (!has_displayed_item_improvements && (craftquality < craftquality_only_imps))
+            craftquality = craftquality_only_imps;
+
+        // First, actual item quality
+        addQuality(tmp, craftquality);
+
+        // Next, magic enchants
+        if (item->getMagic() != NULL)
+            tmp = '\x11' + tmp + '\x10'; // <| |>
+
+        // Next, improvements
+        if (has_displayed_item_improvements) {
             tmp = '\xAE' + tmp + '\xAF'; // («) + tmp + (»)
-            addQuality(tmp, item->getImprovementQuality());
+            if (df::global::d_init->display.flags.is_set(d_init_flags1::SHOW_IMP_QUALITY))
```

Comment:
although we've traditionally tested globals before using them in the core library, I think this function is well out of the way of the code path that might be hit when globals aren't defined, so this is fine.